miniml/compiler/compile.scm

```diff
@@ -1079,41 +1079,172 @@
   (vars switch-block-get-vars)
   (rhs switch-block-get-rhs))
 
-(define (split-pattern-matching env m)
-  (match m
-    (#nil empty-switch)
-    (((p . e) . rest)
-     (match p
-       (('PWild)
-        (switch-with-default empty-switch (cons #nil e)))
-       (('PVar v)
-        (switch-with-default empty-switch (cons v e)))
-       (('PInt n)
-        (let* ((sw-rest (split-pattern-matching env rest))
-               (sw (switch-no-nums sw-rest)))
-          (switch-cons-const sw (mkswitch-const n e))))
-       (('PConstr c l)
-        (let* ((sw-rest (split-pattern-matching env rest))
-               (cdef (env-get-constr env c))
-               (arity (constr-get-arity cdef))
-               (tag (constr-get-tag cdef))
-               (cnums (constr-get-numconstrs cdef))
-               (const (mkswitch-const tag e))
-               (l (adjust-pconstr-args l arity))
-               (vars (map (match-lambda
-                           (('PVar v) v)
-                           (('PWild) #nil)) l))
-               (block (mkswitch-block tag arity vars e))
-               (sw (switch-merge-nums sw-rest cnums)))
-          (match l
-            (#nil
-             (assert (or (= arity 0) (= arity -1)))
-             (switch-cons-const sw const))
-            (('PWild)
-              (switch-cons-block sw block))
-            (_
-             (assert (or (= arity -1) (= arity (length l))))
-             (switch-cons-block sw block)))))))))
+; Pattern-matching compilation is done using "matrix decomposition".
+; A matrix represents an ordered list of pattern-matching clauses that
+; simultaneously match on several variables (rather than arbitrary expressions)
+;
+; The matrix corresponding to
+;   match x,  y,  z with
+;      |  p1, q1, r1 -> e1
+;      |  p2, q2, r2 -> e2
+; is represented by the "arguments" list (args)
+;   (list x y z)
+; paired with the "matrix" of clauses (m)
+;   (list
+;    (list (list p1 q1 r1) e1)
+;    (list (list p2 q2 r2) e2))
+; The "arity" of a matrix is the size of its (args)
+; list, and the number of columns of the matrix --
+; the two quantities always remain equal.
+;
+; Note: matrix rows are represented as fixed-sized lists
+; (one element for the patterns, one for the right-hand-side)
+; instead of cons-pairs, as further specialized representations
+; are used during the transformation at higher arities
+; (3 for non-empty rows, 5 for non-empty rows with beheaded simple first pattern)
+; and cons-pairs quickly become inconvenient at higher arities.
```

Comment:
I am planning to change this to use records instead of fixed-size lists (fixed-size lists look fine but they are used nowhere except in AST structures) at some point.